### PR TITLE
Updates to Dockerfiles

### PR DESCRIPTION
**Pull request by thaJeztah**
Just some updates to the Dockerfiles, as I noticed some things were outdated, and I couldn't stop myself from making some other changes;

- Update to alpine 3.7 for the `.run` version
- Update docker-cli to 18.03.1-ce
- Some optimisations / changes to improve consistency, and optimise cache

There's still some improvements to be made (and it looks like the armhf/s390x versions aren't really used), but didn't want to spend too much time on it for now, haha

### CODE REVIEW

**Comment by thaJeztah** [comment]
@shin- best go through the commits one at a time to follow what I did (and why); also double check that I didn't mess things up (as I'm not really a pythonist, so perhaps there was a specific reason for doing things the old way 😅 )

File: Dockerfile [hunk 1]
```diff
@@ -8,32 +8,32 @@ RUN set -ex; \
...
+    echo "${DOCKERBINS_SHA} dockerbins.tgz" | sha256sum -c - && \
+    tar xvf dockerbins.tgz docker/docker --strip-components 1 && \
+    mv docker /usr/local/bin/docker && \
+    chmod +x /usr/local/bin/docker && \
+    rm dockerbins.tgz
+
 WORKDIR /code/
 
-RUN pip install tox==2.1.1
+COPY --chown=user:user . /code/
```
Comment (shin-) [inline Dockerfile]: The point of doing individual ADDs like below is to avoid having to re-run the tox environment setup when making unrelated changes (like changes to test files). Ideally this is still something we'd want to keep.

**Review by shin- — COMMENTED** [review]
Just one optimization we'd like to keep, otherwise this looks fine 👍 thanks!

**Comment by deivid-rodriguez** [comment]
Great PR for learning `docker`, thanks! :). I guess getting this ready is a matter of partially reverting https://github.com/docker/compose/pull/6061/commits/c183cd3ca117ef227f35ed3c6c1cca52058a1620, and fixing conflicts!

**Comment by ulyssessouza** [comment]
Dropping this PR in favor of https://github.com/docker/compose/pull/6641

Thanks @thaJeztah !
